pkg/kubecost/asset.go

```diff
@@ -2716,80 +2858,68 @@ func (as *AssetSet) ReconciliationMatchMap() map[string]map[string]Asset {
 // Get returns the Asset in the AssetSet at the given key, or nil and false
 // if no Asset exists for the given key
 func (as *AssetSet) Get(key string) (Asset, bool) {
-	as.RLock()
-	defer as.RUnlock()
-
-	if a, ok := as.assets[key]; ok {
+	if a, ok := as.Assets[key]; ok {
 		return a, true
 	}
 	return nil, false
 }
 
 // Insert inserts the given Asset into the AssetSet, using the AssetSet's
-// configured properties to determine the key under which the Asset will
+// configured Properties to determine the key under which the Asset will
 // be inserted.
 func (as *AssetSet) Insert(asset Asset) error {
 	if as == nil {
 		return fmt.Errorf("cannot Insert into nil AssetSet")
 	}
 
-	as.Lock()
-	defer as.Unlock()
-
-	if as.assets == nil {
-		as.assets = map[string]Asset{}
+	if as.Assets == nil {
+		as.Assets = map[string]Asset{}
 	}
 
 	// Determine key into which to Insert the Asset.
-	k, err := key(asset, as.aggregateBy)
+	k, err := key(asset, as.AggregationKeys)
 	if err != nil {
 		return err
 	}
 
 	// Add the given Asset to the existing entry, if there is one;
 	// otherwise just set directly into assets
-	if _, ok := as.assets[k]; !ok {
-		as.assets[k] = asset
+	if _, ok := as.Assets[k]; !ok {
+		as.Assets[k] = asset
+
+		// insert the asset into it's type specific map as well
+		addToConcreteMap(as, k, asset)
 	} else {
-		as.assets[k] = as.assets[k].Add(asset)
+		// possibly creates a new asset type, so we need to update the
+		// concrete type mappings
+		newAsset := as.Assets[k].Add(asset)
+		removeFromConcreteMap(as, k, as.Assets[k])
+
+		// overwrite the existing asset with the new one, and re-add the new
+		// asset to the concrete type mappings
+		as.Assets[k] = newAsset
+		addToConcreteMap(as, k, newAsset)
```

Comment:
> if the key doesn't exist, then set into Assets and addToConcreteMap; my only question here is... will we end up with a Node in the nodeMap? Should we just have an Any, because we're aggregating by e.g. cluster ❓

I'm not sure I follow. If the type we're inserting is a Node, it will definitely end up in the Node map. If you're saying there's a possibility to omit adding/removing the Node when we know that eventually we'll be moving that Node to an Any, I'd prefer the consistent route unless we can always guarantee a specific condition on the first add. 

> One thought -- do you think it's worth writing some extra unit tests to confirm this for a few cases? That's something I'd be up for helping with. Just want to be careful here.

Yeah, I think I have some tests in the bingen side of things that I can move over quickly....